R-package/src/Makevars.in

```diff
@@ -9,6 +9,7 @@ LGB_CPPFLAGS = \
 
 PKG_CPPFLAGS = \
     -I$(PKGROOT)/include \
+    -DFMT_HEADER_ONLY \
```

Comment:
/gha run r-valgrind

Reply:
kicked off this job: https://github.com/microsoft/LightGBM/actions/runs/340857492

Reply:
ok the job got a ✅ but I think that's because the "did this fail" condition is more complicated than I thought in #3443 

@guolinke @StrikerRUS does this new issue look related to this PR?

```text
==1768== Invalid read of size 1
==1768==    at 0x1537F2E3: LightGBM::LeafSplits::Init(float const*, float const*) (leaf_splits.hpp:74)
==1768==    by 0x15373078: LightGBM::SerialTreeLearner::BeforeTrain() (serial_tree_learner.cpp:264)
==1768==    by 0x1537298D: LightGBM::SerialTreeLearner::Train(float const*, float const*) (serial_tree_learner.cpp:166)
==1768==    by 0x151C725C: LightGBM::GBDT::TrainOneIter(float const*, float const*) (gbdt.cpp:379)
==1768==    by 0x154A2DF8: LightGBM::Booster::TrainOneIter(float const*, float const*) (c_api.cpp:364)
==1768==    by 0x1549412A: LGBM_BoosterUpdateOneIterCustom (c_api.cpp:1634)
==1768==    by 0x154B6F4D: LGBM_BoosterUpdateOneIterCustom_R (lightgbm_R.cpp:417)
==1768==    by 0x4941BDA: R_doDotCall (dotcode.c:610)
==1768==    by 0x494CCC6: do_dotcall (dotcode.c:1281)
==1768==    by 0x499FB01: bcEval (eval.c:7078)
==1768==    by 0x498B67F: Rf_eval (eval.c:727)
==1768==    by 0x498E414: R_execClosure (eval.c:1895)
```

Reply:
ok this was not related and was fixed in #3525

Reply:
/gha run r-valgrind

Reply:
that comment kicked off this build, let's see how it goes: https://github.com/microsoft/LightGBM/actions/runs/355061507

Reply:
I also just started Solaris 10 builds on R Hub, hopefully they go well!

```r
result <- rhub::check(
    path = "lightgbm_3.0.0.99.tar.gz"
    , email = "jaylamb20@gmail.com"
    , check_args = c(
        "--as-cran"
    )
    , platform = c(
        "solaris-x86-patched"
        , "solaris-x86-patched-ods"
    )
    , env_vars = c(
        "R_COMPILE_AND_INSTALL_PACKAGES" = "always"
        , "_R_CHECK_FORCE_SUGGESTS_" = "true"
        , "_R_CHECK_CRAN_INCOMING_USE_ASPELL_" = "true"
    )
)
```

Reply:
Unfortunately, the current state of this branch would result in the R package being rejected from CRAN, because it seems `fast_double_parser` cannot be built on Solaris 10.

```text
In file included from ./include/LightGBM/utils/common.h:32:0,
                 from ./include/LightGBM/config.h:16,
                 from ./include/LightGBM/application.h:8,
                 from application/application.cpp:5:
./include/LightGBM/fast_double_parser.h: In function ‘const char* fast_double_parser::parse_float_strtod(const char*, double*)’:
./include/LightGBM/fast_double_parser.h:1096:10: error: ‘locale_t’ does not name a type
   static locale_t c_locale = newlocale(LC_ALL_MASK, "C", NULL);
          ^
./include/LightGBM/fast_double_parser.h:1097:39: error: ‘c_locale’ was not declared in this scope
   *outDouble = strtod_l(ptr, &endptr, c_locale);
                                       ^
./include/LightGBM/fast_double_parser.h:1097:47: error: ‘strtod_l’ was not declared in this scope
   *outDouble = strtod_l(ptr, &endptr, c_locale);
                                               ^
gmake: *** [/opt/csw/lib/R/etc/Makeconf:181: application/application.o] Error 1
ERROR: compilation failed for package ‘lightgbm’
* removing ‘/export/home/XaoAKgd/R/lightgbm’
Warning message:
In i.p(...) :
  installation of package ‘/export/home/XaoAKgd/Rtemp/RtmpQdrjZj/file5af04b2d6989/lightgbm_3.0.0.99.tar.gz’ had non-zero exit status
> 
> 
```

If it helps, I found the following possibly-related discussions:

* https://github.com/fontforge/fontforge/issues/3027#issuecomment-279061231
* https://github.com/fmtlib/fmt/issues/1240

@lemire is this something you could help with? I think it's something that would have to be fixed for `fast_double_parser` to be used in any R package on CRAN.

<details><summary>how to run tests on Solaris</summary>

From the root of this repo

```shell
git submodule init
git submodule update
docker run \
    -v $(pwd):/tmp/LightGBM \
    -w /tmp/LightGBM \
    -it rhub/rocker-gcc-san \
     /bin/bash
```

Then, in the shell that prompts up

```shell
sh build-cran-package.sh

# start an R repl
R
```

```R
install.packages("rhub", repos = "https://cran.r-project.org")

#  put in your email
result <- rhub::check(
    path = "lightgbm_3.0.0.99.tar.gz"
    , email =  YOUR-EMAIL-IN-QUOTES-HERE
    , check_args = c(
        "--as-cran"
    )
    , platform = c(
        "solaris-x86-patched"
        , "solaris-x86-patched-ods"
    )
    , env_vars = c(
        "R_COMPILE_AND_INSTALL_PACKAGES" = "always"
        , "_R_CHECK_FORCE_SUGGESTS_" = "true"
        , "_R_CHECK_CRAN_INCOMING_USE_ASPELL_" = "true"
    )
)
```

</details>

Reply:
but `valgrind` is happy still! 🤗 

https://github.com/microsoft/LightGBM/runs/1377534620?check_suite_focus=true